File: Tutorial5-3aDagger-Android/src/main/java/com/example/tutorial5_1dagger_android/di/ToastMakerSubComponent.java
package com.example.tutorial5_1dagger_android.di;


import android.content.Context;

import com.example.tutorial5_1dagger_android.MainActivity;

import dagger.BindsInstance;
import dagger.Subcomponent;
import dagger.android.AndroidInjector;

@ActivityScope
@Subcomponent(modules = {ToastMakerModule.class})
public interface ToastMakerSubComponent extends AndroidInjector<MainActivity> {


    @Subcomponent.Builder
    abstract class Builder extends AndroidInjector.Builder<MainActivity> {

        // It gets concrete type instance from AndroidInjector<T>

        /*
            These 2 methods get Context from MainActivity
          */
        @Override
        public void seedInstance(MainActivity instance) {
            context(instance);
        }

        @BindsInstance
        public abstract Builder context(Context context);
    }

}

File: Tutorial5-2Dagger-Android-SubComponents2SeedInstance/src/main/java/com/example/tutorial5_2dagger_android_subcomponents2/di/ToastMakerSubComponent.kt
package com.example.tutorial5_2dagger_android_subcomponents2.di

import android.content.Context
import com.example.tutorial5_2dagger_android_subcomponents2.MainActivity
import dagger.BindsInstance
import dagger.Subcomponent
import dagger.android.AndroidInjector

/**
 * SubComponents extend AndroidInjector<MainActivity> for scoped injections.
 * [MainActivity] is the class where [ToastMakerModule] objects are injected.
 *
 *
 *
 * Sub component and @Provides annotated methods inside module CAN NOT not have different scopes.
 *
 *
 * IMPORTANT: Both can be un-scoped but cannot have with different scopes.
</MainActivity> */
@ActivityScope
@Subcomponent(modules = [ToastMakerModule::class])
interface ToastMakerSubComponent : AndroidInjector<MainActivity> {
   
    @Subcomponent.Builder
    abstract class Builder : AndroidInjector.Builder<MainActivity>() {

        // It gets concrete type instance from AndroidInjector<T>
        /*
          These 2 methods get Context from MainActivity
        */
        override fun seedInstance(instance: MainActivity) {
            context(instance)
        }

        @BindsInstance
        abstract fun context(context: Context): Builder
    }
}